### PR TITLE
Reduce memory usage for fuse mount

### PR DESCRIPTION
**Pull request by fd0**
This changes `repository.LoadBlob()` so that a destination buffer must be provided, which enables the fuse code to use a buffer from a `sync.Pool`. In addition, release the buffers when the file is closed. At the moment, the max memory usage is defined by the max file size that is read in one go (e.g. with `cat`). It could be further optimized by implementing a LRU caching scheme.

Also refactor the fuse `Read` method a bit.

Connects #248.

### CODE REVIEW

**Comment by fd0** [comment]
I think we also need tests for all the fuse stuff...

**Comment by fd0** [comment]
I felt very unsure about the magic happening in `file.Read()`, so I refactored this a bit to be able to mock the repository, and added tests for `Read()`.

File: cmd/restic/cmd_cat.go [hunk 1]
```diff
@@ -162,9 +162,18 @@ func (cmd CmdCat) Execute(args []string) error {
 		return err
 
 	case "blob":
-		data, err := repo.LoadBlob(pack.Data, id)
-		if err == nil {
-			_, err = os.Stdout.Write(data)
+		_, tpe, _, length, err := repo.Index().Lookup(id)
```
Comment (fw42) [inline cmd/restic/cmd_cat.go]: "tpe"? what is a "tpe"? maybe "blobType"?

File: repository/repository.go [hunk 2]
```diff
@@ -97,6 +97,10 @@ func (r *Repository) LoadBlob(t pack.BlobType, id backend.ID) ([]byte, error) {
 		return nil, err
 	}
 
+	if length > uint(cap(buf))+crypto.Extension {
+		return nil, errors.New("buf is too small")
```
Comment (fw42) [inline repository/repository.go]: maybe include the number of bytes it needs to be able to hold, to make the error message more useful?

File: repository/repository.go [hunk 3]
```diff
@@ -111,7 +115,9 @@ func (r *Repository) LoadBlob(t pack.BlobType, id backend.ID) ([]byte, error) {
 		return nil, err
 	}
 
-	buf, err := ioutil.ReadAll(rd)
+	// make buffer that is large enough for the complete blob
+	cbuf := make([]byte, length)
```
Comment (fw42) [inline repository/repository.go]: can you think of a more descriptive variable name than `cbuf`? I'm assuming this is a buffer that holds ciphertext? how about `ciphertextBuf`?

File: repository/repository.go [hunk 4]
```diff
@@ -122,17 +128,17 @@ func (r *Repository) LoadBlob(t pack.BlobType, id backend.ID) ([]byte, error) {
 	}
 
 	// decrypt
-	plain, err := r.Decrypt(buf)
+	buf, err = r.decryptTo(buf, cbuf)
```
Comment (fw42) [inline repository/repository.go]: and `plaintextBuf` for consistency then? I think that would make this code a bit easier to follow

File: cmd/restic/cmd_cat.go (hunk 1 above)
Comment (fd0) [inline cmd/restic/cmd_cat.go]: Yes, that's better.

File: repository/repository.go [hunk 5]
```diff
@@ -87,8 +86,9 @@ func (r *Repository) LoadAndDecrypt(t backend.Type, id backend.ID) ([]byte, erro
 }
 
 // LoadBlob tries to load and decrypt content identified by t and id from a
-// pack from the backend.
-func (r *Repository) LoadBlob(t pack.BlobType, id backend.ID) ([]byte, error) {
+// pack from the backend, the result is stored in buf, which must be large
+// enough to hold the complete blob.
+func (r *Repository) LoadBlob(t pack.BlobType, id backend.ID, buf []byte) ([]byte, error) {
```
Comment (fw42) [inline repository/repository.go]: what's the reasoning for this method still returning the buffer? Isn't that unnecessary now that it's a parameter? The one that will be returned is always the same as the one that was passed, right? (except for error cases)

File: repository/repository.go (hunk 2 above)
Comment (fd0) [inline repository/repository.go]: okay

File: repository/repository.go (hunk 3 above)
Comment (fd0) [inline repository/repository.go]: Indeed, it is.

**Comment by fw42** [comment]
some bitching, but looks mostly good :-) :+1:

File: repository/repository.go (hunk 5 above)
Comment (fd0) [inline repository/repository.go]: That's correct, but `len` of `buf` could've been modified (and this happens). A slice in Go is call-by-value regarding `len`, but call-by-reference regarding the underlying array. You can think of it as `struct {len int, cap int, data *[200]byte}`, ie. the struct is passed by value, but `data` is a pointer.

So we could either explicitely return `n` (like `Read`), or return the buf slice back to the caller.

**Pull request merged**